source-code/core/src/lint/RFC.md

```diff
@@ -0,0 +1,128 @@
+
+<!------------------------------------------------------------------------------------------------>
+
+## Types
+
+1. I have investigated if [this](https://github.com/orgs/inlang/discussions/319#discussioncomment-4850815) is possible and I could not make it work.
+
+	I have tried 3 different ways to do it. Take a look at the [`experiment-1.ts`](https://github.com/ivanhofer/inlang/blob/lint/source-code/core/src/lint/experiment-1.ts) file to see the examples. I find variant 3 to be the best because it does not hide the properties behind an abstraction (see screenshots below). It can be easily extended in the future with additional properties.
+
+2. We need to extend the Resource type on each level (`Resource`, `Message`, `Pattern`) because each level could have lint information attached to it.
+
+	- "missing key" would be on the `Resource` level
+	- "missing parameter" would be on the `Message` level
+	- "unknown parameter" would be on the `Pattern` level
+
+	Maybe `Pattern` requires another level in the future, but if we solve the problem for the first 3 levels, adding the 4th should be easy.
+
+	Kind of similar to the `metadata` property which can be present on all levels. If we solve the problem for `lint`, we can get rid of the non-typed `metadata` property and plugin authors can use a slightly different approach (additional properties should be scoped to a "namespace"-like object). The result would be a `Resource` with correctly typed `metadata` property.
+
+	See [`experiment-2.ts`](https://github.com/ivanhofer/inlang/blob/lint/source-code/core/src/lint/experiment-2.ts) for an example how this could be done. The only downside is, that those generics make the code harder to read. And VS Code will show jibberish when hovering over a variable (see screenshots below). But the only alternative I can see is to define each Permutation manually which is even worse.
+
+3. What information would a lint error emit?
+
+	```ts
+	type LintInformation = {
+		type: LintRule // which rule emitted the error
+		message: string // a string that informs what is wrong
+	}
+	```
+	Question: Anything else?
+
+	Question: Should any lint rule be able to extend this object?
+
+<!------------------------------------------------------------------------------------------------>
+
+## Configure Lint behavior
+
+`inlang` will provide some lint functionality that can be configured or disabled.
+It does not need to be as complicated as eslint, but it should follow the same design when it comes to the `rules`.
+The `inlang.config.js` will be extended with a new `linting` property.
+
+```js
+/**
+ * @type {import("@inlang/core/config").DefineConfig}
+ */
+export async function defineConfig(env) {
+  	return {
+   	referenceLanguage: 'en,
+   	languages: ['en', 'de'],
+		linting: {
+			missing_key: false, // disable this specific lint rule
+			missing_key: 'warn', // override the level
```

Comment:
This seems to be a good approach 👍
So no built-in rules and instead a plugin that can be independently updated, versioned and therefore can potentially contain breaking changes.
Just the framework for now and a dedicated `@inlang/lint-plugin` package with the `missingKey` rule.

Because each plugin needs to be added individually it is also great to get proper `TypeScript` support becasue each plugin can provide them individually.
```js
{
	lint: [
		inlangLintPlugin({
			missingKey: false, // disable this specific lint rule
		}),
		// ... another plugin
	]
}
```
Plugins need to be `JavaScript` functions. I would not do the same as `eslint` does where you specify a `string` and it tries to resolve the package name and import the npm module. This is the most confusing part for me when using `eslint` (and not having proper `TypeScript` support when defining rules).

Question: the `lint` config just needs to be an array of plugins. Or do we need something else?
Or should we nest it?
```js
{
	lint: {
		plugins: [
			inlangLintPlugin({

			})
		]
	}
}
```

We probably need to mark the error result now to know where it comes from, because any package could implement a `missingKey` lint rule.

```diff
export type LintInformation = {
+    plugin: string
	type: LintRule
	message: string
}
```

Reply:
> Just the framework for now and a dedicated @inlang/lint-plugin package with the missingKey rule. 

Yes. I would name the lint-plugin `@inlang/standard-lint-rules-plugin` or similar to clearly communicate that this is an optional plugin. `@inlang/lint-plugin` could be understood as "I need this plugin to get lints". 

> Plugins need to be JavaScript functions. I would not do the same as eslint does where you specify a string and it tries to resolve the package name and import the npm module. This is the most confusing part for me when using eslint (and not having proper TypeScript support when defining rules). 

Agree. Eslint has been designed with JSON as a config format in mind. We use JavaScript, so direct imports (and more) are possible. We should leverage that. 

> Question: the lint config just needs to be an array of plugins. Or do we need something else?

Let's try to keep the API surface as small as possible while allowing us to extend the schema in the future with low risk of breaking changes.

I proposed the following syntax [here](https://github.com/inlang/inlang/pull/351#discussion_r1096556538):

 ```ts

{
  lint: {
    rules: [
      standardRules(),
      germanExtraRules(), 
      sentimentRules({
          // custom rule config
          minimumScore: "0.3"
      })
   ]
  }
}

```

The question is what's the input/output API for a "rule function"?  

> Or should we nest it? 

Do you refer to having a `lint: Array<RuleFunction>` or `lint: { rules: Array<RuleFunction> }`? 

I strongly vote for the latter. The 'nested' config allows us to extend the lint property in the future with, for example, a `disable` property etc.


> We probably need to mark the error result now to know where it comes from, because any package could implement a missingKey lint rule.

I would give each Lint an id. Plugin authors should prefix the id with their plugin name. 

export type LintInformation = {
+      id: string
	type: LintRule (error | warning)
	message: string
}

Reply:
> Yes. I would name the lint-plugin `@inlang/standard-lint-rules-plugin`

👍

> Let's try to keep the API surface as small as possible while allowing us to extend the schema in the future with low risk of breaking changes

Ok, then let's stick to a nested `rules` property.

> The question is what's the input/output API for a "rule function"?

I think we need to first implement a rule to see what makes sense.

> I would give each Lint an id. Plugin authors should prefix the id with their plugin name.

The prefix is probably a better option because then the id property is really unique. and we don't need to check two properties to see if it is the rule of a certain plugin.

Reply:
> I think we need to first implement a rule to see what makes sense.

Agree. Your `experiment-x.ts` approach is great. Could be used for testing rules API's too.